File: services/sensor-service/service/sensor_test.go
package service

import (
	"context"
	"database/sql/driver"
	"errors"
	"testing"

	"github.com/go-kit/kit/log"
	"github.com/opentracing/opentracing-go"
	"github.com/opentracing/opentracing-go/mocktracer"
	"github.com/stretchr/testify/assert"
	"gopkg.in/DATA-DOG/go-sqlmock.v1"
)

func CreateContextWithSpan() context.Context {
	tracer := mocktracer.New()
	span := tracer.StartSpan("mock-span")
	ctx := opentracing.ContextWithSpan(context.Background(), span)
	return ctx
}

func TestNewSensorManager(t *testing.T) {
	tests := []struct {
		name        string
		postgresURL string
	}{
		{
			"WithoutUserPass",
			"postgres://localhost",
		},
		{
			"WithUserPass",
			"postgres://root:pass@localhost",
		},
	}

	for _, tc := range tests {
		t.Run(tc.name, func(t *testing.T) {
			db := NewPostgresDB(tc.postgresURL)
			logger := log.NewNopLogger()
			tracer := mocktracer.New()
			m := NewSensorManager(db, logger, tracer)

			assert.NotNil(t, m)
		})
	}
}

func TestSensorManagerCreate(t *testing.T) {
	tests := []struct {
		name          string
		dbError       error
		dbResult      driver.Result
		context       context.Context
		sensorSiteID  string
		sensorName    string
		sensorUnit    string
		sensorMinSafe float64
		sensorMaxSafe float64
		expectError   bool
	}{
		{
			"DatabaseError",
			errors.New("db error"),
			nil,
			CreateContextWithSpan(),
			"", "", "", 0.0, 0.0,
			true,
		},
		{
			"CreateSensor",
			nil,
			sqlmock.NewResult(0, 1),
			CreateContextWithSpan(),
			"1111-aaaa", "temperature", "celsius", -30.0, 30.0,
			false,
		},
	}

	for _, tc := range tests {
		t.Run(tc.name, func(t *testing.T) {
			db, mock, err := sqlmock.New()
			assert.NoError(t, err)

			m := &postgresSensorManager{
				db:     db,
				logger: log.NewNopLogger(),
				tracer: mocktracer.New(),
			}

			// Mock SQL query
			expect := mock.ExpectExec(`INSERT INTO sensors`)
			if tc.dbError != nil {
				expect.WillReturnError(tc.dbError)
			} else {
				expect.WillReturnResult(tc.dbResult)
			}

			sensor, err := m.Create(tc.context, tc.sensorSiteID, tc.sensorName, tc.sensorUnit, tc.sensorMinSafe, tc.sensorMaxSafe)

			if tc.expectError {
				assert.Error(t, err)
				assert.Nil(t, sensor)
			} else {
				assert.NoError(t, err)
				assert.NotEmpty(t, sensor.ID)
				assert.Equal(t, tc.sensorSiteID, sensor.SiteID)
				assert.Equal(t, tc.sensorName, sensor.Name)
				assert.Equal(t, tc.sensorUnit, sensor.Unit)
				assert.Equal(t, tc.sensorMinSafe, sensor.MinSafe)
				assert.Equal(t, tc.sensorMaxSafe, sensor.MaxSafe)
			}
		})
	}
}

func TestSensorManagerAll(t *testing.T) {
	tests := []struct {
		name            string
		dbError         error
		dbRows          [][]driver.Value
		context         context.Context
		siteID          string
		expectError     bool
		expectedSensors []Sensor
	}{
		{
			"DatabaseError",
			errors.New("db error"),
			nil,
			CreateContextWithSpan(),
			"",
			true,
			nil,
		},
		{
			"AllSensors",
			nil,
			[][]driver.Value{
				[]driver.Value{"2222-bbbb", "1111-aaaa", "temperature", "fahrenheit", -22.0, 86.0},
				[]driver.Value{"3333-cccc", "1111-aaaa", "pressure", "pascal", 50000.0, 100000.0},
			},
			CreateContextWithSpan(),
			"1111-aaaa",
			false,
			[]Sensor{
				Sensor{"2222-bbbb", "1111-aaaa", "temperature", "fahrenheit", -22.0, 86.0},
				Sensor{"3333-cccc", "1111-aaaa", "pressure", "pascal", 50000.0, 100000.0},
			},
		},
	}

	for _, tc := range tests {
		t.Run(tc.name, func(t *testing.T) {
			db, mock, err := sqlmock.New()
			assert.NoError(t, err)

			m := &postgresSensorManager{
				db:     db,
				logger: log.NewNopLogger(),
				tracer: mocktracer.New(),
			}

			// Mock SQL query
			expect := mock.ExpectQuery(`SELECT id, site_id, name, unit, min_safe, max_safe FROM sensors`)
			if tc.dbError != nil {
				expect.WillReturnError(tc.dbError)
			} else {
				rows := sqlmock.NewRows([]string{"id", "site_id", "name", "unit", "min_safe", "max_safe"})
				for _, row := range tc.dbRows {
					rows.AddRow(row...)
				}
				expect.WillReturnRows(rows)
			}

			sensors, err := m.All(tc.context, tc.siteID)

			if tc.expectError {
				assert.Error(t, err)
				assert.Nil(t, sensors)
			} else {
				assert.NoError(t, err)
				assert.NotNil(t, sensors)

				for i, sensor := range sensors {
					assert.Equal(t, tc.expectedSensors[i].ID, sensor.ID)
					assert.Equal(t, tc.expectedSensors[i].SiteID, sensor.SiteID)
					assert.Equal(t, tc.expectedSensors[i].Name, sensor.Name)
					assert.Equal(t, tc.expectedSensors[i].Unit, sensor.Unit)
					assert.Equal(t, tc.expectedSensors[i].MinSafe, sensor.MinSafe)
					assert.Equal(t, tc.expectedSensors[i].MaxSafe, sensor.MaxSafe)
				}
			}
		})
	}
}

func TestSensorManagerGet(t *testing.T) {
	tests := []struct {
		name           string
		dbError        error
		dbRow          []driver.Value
		context        context.Context
		id             string
		expectError    bool
		expectedSensor *Sensor
	}{
		{
			"DatabaseError",
			errors.New("db error"),
			nil,
			CreateContextWithSpan(),
			"",
			true,
			nil,
		},
		{
			"GetSensor",
			nil,
			[]driver.Value{"2222-bbbb", "1111-aaaa", "temperature", "fahrenheit", -22.0, 86.0},
			CreateContextWithSpan(),
			"2222-bbbb",
			false,
			&Sensor{"2222-bbbb", "1111-aaaa", "temperature", "fahrenheit", -22.0, 86.0},
		},
	}

	for _, tc := range tests {
		t.Run(tc.name, func(t *testing.T) {
			db, mock, err := sqlmock.New()
			assert.NoError(t, err)

			m := &postgresSensorManager{
				db:     db,
				logger: log.NewNopLogger(),
				tracer: mocktracer.New(),
			}

			// Mock SQL query
			expect := mock.ExpectQuery(`SELECT id, site_id, name, unit, min_safe, max_safe FROM sensors`)
			if tc.dbError != nil {
				expect.WillReturnError(tc.dbError)
			} else {
				rows := sqlmock.NewRows([]string{"id", "site_id", "name", "unit", "min_safe", "max_safe"})
				rows.AddRow(tc.dbRow...)
				expect.WillReturnRows(rows)
			}

			sensor, err := m.Get(tc.context, tc.id)

			if tc.expectError {
				assert.Error(t, err)
				assert.Nil(t, sensor)
			} else {
				assert.NoError(t, err)
				assert.Equal(t, tc.expectedSensor.ID, sensor.ID)
				assert.Equal(t, tc.expectedSensor.SiteID, sensor.SiteID)
				assert.Equal(t, tc.expectedSensor.Name, sensor.Name)
				assert.Equal(t, tc.expectedSensor.Unit, sensor.Unit)
				assert.Equal(t, tc.expectedSensor.MinSafe, sensor.MinSafe)
				assert.Equal(t, tc.expectedSensor.MaxSafe, sensor.MaxSafe)
			}
		})
	}
}

func TestSensorManagerUpdate(t *testing.T) {
	tests := []struct {
		name               string
		dbError            error
		dbResult           driver.Result
		context            context.Context
		sensor             Sensor
		expectError        bool
		expectedAffectedNo int
	}{
		{
			"DatabaseError",
			errors.New("db error"),
			nil,
			CreateContextWithSpan(),
			Sensor{"", "", "", "", 0.0, 0.0},
			true,
			0,
		},
		{
			"UpdateSensor",
			nil,
			sqlmock.NewResult(0, 1),
			CreateContextWithSpan(),
			Sensor{"2222-bbbb", "1111-aaaa", "temperature", "fahrenheit", -22.0, 86.0},
			false,
			1,
		},
	}

	for _, tc := range tests {
		t.Run(tc.name, func(t *testing.T) {
			db, mock, err := sqlmock.New()
			assert.NoError(t, err)

			m := &postgresSensorManager{
				db:     db,
				logger: log.NewNopLogger(),
				tracer: mocktracer.New(),
			}

			// Mock SQL query
			expect := mock.ExpectExec(`UPDATE sensors SET`)
			if tc.dbError != nil {
				expect.WillReturnError(tc.dbError)
			} else {
				expect.WillReturnResult(tc.dbResult)
			}

			n, err := m.Update(tc.context, tc.sensor)

			if tc.expectError {
				assert.Error(t, err)
				assert.Zero(t, n)
			} else {
				assert.NoError(t, err)
				assert.Equal(t, tc.expectedAffectedNo, n)
			}
		})
	}
}

func TestSensorManagerDelete(t *testing.T) {
	tests := []struct {
		name        string
		dbError     error
		dbResult    driver.Result
		context     context.Context
		sensorID    string
		expectError bool
	}{
		{
			"DatabaseError",
			errors.New("db error"),
			nil,
			CreateContextWithSpan(),
			"",
			true,
		},
		{
			"DeleteSensor",
			nil,
			sqlmock.NewResult(0, 1),
			CreateContextWithSpan(),
			"2222-bbbb",
			false,
		},
	}

	for _, tc := range tests {
		t.Run(tc.name, func(t *testing.T) {
			db, mock, err := sqlmock.New()
			assert.NoError(t, err)

			m := &postgresSensorManager{
				db:     db,
				logger: log.NewNopLogger(),
				tracer: mocktracer.New(),
			}

			// Mock SQL query
			expect := mock.ExpectExec(`DELETE FROM sensors`)
			if tc.dbError != nil {
				expect.WillReturnError(tc.dbError)
			} else {
				expect.WillReturnResult(tc.dbResult)
			}

			err = m.Delete(tc.context, tc.sensorID)

			if tc.expectError {
				assert.Error(t, err)
			} else {
				assert.NoError(t, err)
			}
		})
	}
}

File: services/graphql-server/src/graphql/resolvers/sensor.test.js
/* eslint-env mocha */
const sinon = require('sinon')
const should = require('should')

const resolvers = require('./sensor')

describe('sensorResolvers', () => {
  let span, logger
  let sensorService, _sensorService
  let context, info

  beforeEach(() => {
    span = {}
    logger = {
      debug () {},
      verbose () {},
      info () {},
      warn () {},
      error () {},
      fatal () {}
    }

    sensorService = {
      get () {},
      all () {},
      create () {},
      update () {},
      delete () {},
    }
    _sensorService = sinon.mock(sensorService)

    context = { span, logger, sensorService }
    info = {}
  })

  afterEach(() => {
    _sensorService.restore()
  })

  describe('Query', () => {
    describe('sensor', () => {
      let id

      it('should throw an error when service.get fails', done => {
        id = '1111-1111'
        const err = new Error("get error")
        _sensorService.expects('get').withArgs({ span }, id).rejects(err)
        resolvers.Query.sensor(null, { id }, context, info).catch(e => {
          e.should.eql(err)
          _sensorService.verify()
          done()
        })
      })
      it('should return a promise that resolves to a sensor', done => {
        id = '2222-2222'
        const sensor = { id: '1111-1111', siteId: 'aaaa-aaaa', name: 'temperature', unit: 'celsius', minSafe: -30.0, maxSafe: 30.0 }
        _sensorService.expects('get').withArgs({ span }, id).resolves(sensor)
        resolvers.Query.sensor(null, { id }, context, info).then(s => {
          s.should.eql(sensor)
          _sensorService.verify()
          done()
        }).catch(done)
      })
    })

    describe('sensors', () => {
      let siteId

      it('should throw an error when service.all fails', done => {
        siteId = 'aaaa-aaaa'
        const err = new Error("all error")
        _sensorService.expects('all').withArgs({ span }, siteId).rejects(err)
        resolvers.Query.sensors(null, { siteId }, context, info).catch(e => {
          e.should.eql(err)
          _sensorService.verify()
          done()
        })
      })
      it('should return a promise that resolves to an array sensors', done => {
        siteId = 'bbbb-bbbb'
        const sensors = [{ id: '1111-1111', siteId: 'aaaa-aaaa', name: 'temperature', unit: 'celsius', minSafe: -30.0, maxSafe: 30.0 }]
        _sensorService.expects('all').withArgs({ span }, siteId).resolves(sensors)
        resolvers.Query.sensors(null, { siteId }, context, info).then(s => {
          s.should.eql(sensors)
          _sensorService.verify()
          done()
        }).catch(done)
      })
    })
  })

  describe('Mutation', () => {
    describe('createSensor', () => {
      let input

      it('should throw an error when service.create fails', done => {
        input = {}
        const err = new Error("create error")
        _sensorService.expects('create').withArgs({ span }, input).rejects(err)
        resolvers.Mutation.createSensor(null, { input }, context, info).catch(e => {
          e.should.eql(err)
          _sensorService.verify()
          done()
        })
      })
      it('should return a promise that resolves to the new sensor', done => {
        input = { siteId: 'aaaa-aaaa', name: 'temperature', unit: 'celsius', minSafe: -30.0, maxSafe: 30.0 }
        const sensor = Object.assign({}, { id: '2222-2222' }, input)
        _sensorService.expects('create').withArgs({ span }, input).resolves(sensor)
        resolvers.Mutation.createSensor(null, { input }, context, info).then(s => {
          s.should.eql(sensor)
          _sensorService.verify()
          done()
        }).catch(done)
      })
    })

    describe('updateSensor', () => {
      let id, input

      it('should throw an error when service.update fails', done => {
        id = '1111-1111'
        input = {}
        const err = new Error("update error")
        _sensorService.expects('update').withArgs({ span }, id, input).rejects(err)
        resolvers.Mutation.updateSensor(null, { id, input }, context, info).catch(e => {
          e.should.eql(err)
          _sensorService.verify()
          done()
        })
      })
      it('should return a promise that resolves to the updated sensor', done => {
        id = '2222-2222'
        input = { siteId: 'aaaa-aaaa', name: 'temperature', unit: 'celsius', minSafe: -30.0, maxSafe: 30.0 }
        const sensor = Object.assign({}, { id }, input)
        _sensorService.expects('update').withArgs({ span }, id, input).resolves(sensor)
        resolvers.Mutation.updateSensor(null, { id, input }, context, info).then(s => {
          s.should.eql(sensor)
          _sensorService.verify()
          done()
        }).catch(done)
      })
    })

    describe('deleteSensor', () => {
      let input

      it('should throw an error when service.delete fails', done => {
        id = '1111-1111'
        const err = new Error("delete error")
        _sensorService.expects('delete').withArgs({ span }, id).rejects(err)
        resolvers.Mutation.deleteSensor(null, { id }, context, info).catch(e => {
          e.should.eql(err)
          _sensorService.verify()
          done()
        })
      })
      it('should return a promise that resolves true', done => {
        id = '2222-2222'
        _sensorService.expects('delete').withArgs({ span }, id).resolves(true)
        resolvers.Mutation.deleteSensor(null, { id }, context, info).then(result => {
          result.should.be.true()
          _sensorService.verify()
          done()
        }).catch(done)
      })
    })
  })

  describe('Sensor', () => {
    let sensor

    beforeEach(() => {
      sensor = { id: '1111-1111', siteId: 'aaaa-aaaa', name: 'temperature', unit: 'celsius', minSafe: -30.0, maxSafe: 30.0 }
    })

    describe('id', () => {
      it('should return sensor id', () => {
        const id = resolvers.Sensor.id(sensor)
        should.equal(id, sensor.id)
      })
    })
    describe('siteId', () => {
      it('should return sensor siteId', () => {
        const siteId = resolvers.Sensor.siteId(sensor)
        should.equal(siteId, sensor.siteId)
      })
    })
    describe('name', () => {
      it('should return sensor name', () => {
        const name = resolvers.Sensor.name(sensor)
        should.equal(name, sensor.name)
      })
    })
    describe('unit', () => {
      it('should return sensor unit', () => {
        const unit = resolvers.Sensor.unit(sensor)
        should.equal(unit, sensor.unit)
      })
    })
    describe('minSafe', () => {
      it('should return sensor minSafe', () => {
        const minSafe = resolvers.Sensor.minSafe(sensor)
        should.equal(minSafe, sensor.minSafe)
      })
    })
    describe('maxSafe', () => {
      it('should return sensor maxSafe', () => {
        const maxSafe = resolvers.Sensor.maxSafe(sensor)
        should.equal(maxSafe, sensor.maxSafe)
      })
    })
  })
})
